dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,695 @@
+    pub let MinterPrivatePath:      PrivatePath
+
+    // Entity Counts
+    //
+    pub var totalSupply:    UInt64
+    pub var nextItemID:     UInt64
+
+    // Lists in contract
+    //
+    access(self) let collectionGroupByID: @{UInt64: CollectionGroup}
+    access(self) let slotByID: @{UInt64: Slot}
+    access(self) let itemByID: {UInt64: Item}
+
+    // A public struct to access Item data
+    //
+    pub struct Item {
+        pub let id: UInt64 // unique uuid of resource
+        pub let itemID: UInt64 // the id of the edition, tier, play
+        pub let points: UInt64 // points for item
+        pub let itemType: String // (edition.id, edition.tier, play.id)
+
+        init (
+            id: UInt64,
+            itemID: UInt64,
+            points: UInt64,
+            itemType: String
+        ) {
+            self.id = id
+            self.itemID = itemID
+            self.points = points
+            self.itemType = itemType
+
+            DSSCollection.nextItemID = self.id + 1 as UInt64
+
+            emit ItemCreated(
+                id: self.id,
+                itemID: self.itemID,
+                points: self.points,
+                itemType: self.itemType
+            )
+        }
+    }
+
+    // A public struct to access Slot data
+    //
+    pub struct SlotData {
+        pub let id: UInt64
+        pub let collectionGroupID: UInt64
+        pub let logicalOperator: String // (AND / OR)
+        pub let typeName: String // (A.contractAddress.NFT...)
+        pub var items: [Item]
+
+        init (id: UInt64) {
+            if let slot = &DSSCollection.slotByID[id] as &DSSCollection.Slot? {
+                self.id = slot.id
+                self.collectionGroupID = slot.collectionGroupID
+                self.logicalOperator = slot.logicalOperator
+                self.typeName = slot.typeName
+                self.items = slot.items
+            } else {
+                panic("slot does not exist")
+            }
+        }
+    }
+
+    // A top-level Slot with a unique ID
+    //
+    pub resource Slot {
+        pub let id: UInt64
+        pub let collectionGroupID: UInt64
+        pub let logicalOperator: String // (AND / OR)
+        pub let typeName: String // (A.contractAddress.NFT...)
+        pub var items: [Item]
+
+        // Add item to slot
+        //
+        access(contract) fun addItemToSlot(id: UInt64) {
+            pre {
+                DSSCollection.CollectionGroupData(
+                    id: self.collectionGroupID
+                ).open == true: "collection group not open"
+            }
+
+            let item = DSSCollection.getItem(id: id)
+            self.items.append(item)
+
+            emit ItemAddedToSlot(
+                 id: item.id,
+                 slotID: self.id,
+                 collectionGroupID: self.collectionGroupID
+             )
+        }
+
+        init (
+            collectionGroupID: UInt64,
+            logicalOperator: String,
```

Comment:
We can at least do a check that logicalOperator either should be "AND" / "OR" in pre statement otherwise, there are high changes the admin can mess up.

Reply:
We'll be submitting in through an API, the API will check, but I do think it can't hurt to verify AND/OR here. I don't think it would be too breaking if we added more later, though I'm not sure what we would add (maybe multiply or something like for each of these that it does have you get more points or something idk)

Reply:
I added validation logic to this. Will reach out to Satyam on using a type here.